File: infra/hetzner/phase2-tail/verify-staging-key-role.mjs
#!/usr/bin/env node
/** Prints only JWT role + key length — never the key. */
import fs from "node:fs";

const file = process.argv[2] || "/opt/souq-arab/config/api.env.staging";
const text = fs.readFileSync(file, "utf8");
const m = text.match(/^SUPABASE_SERVICE_ROLE_KEY=(.*)$/m);
if (!m) {
  console.log("role=missing len=0");
  process.exit(1);
}
const tok = m[1].trim().replace(/^["']|["']$/g, "");
const parts = tok.split(".");
if (parts.length < 2) {
  console.log(`role=invalid_jwt len=${tok.length}`);
  process.exit(1);
}
const payloadPart = parts[1];
const decoders = [
  () => Buffer.from(payloadPart, "base64url"),
  () => Buffer.from(payloadPart.replace(/-/g, "+").replace(/_/g, "/"), "base64"),
];
for (const decode of decoders) {
  try {
    const payload = JSON.parse(decode().toString("utf8"));
    console.log(`role=${payload.role ?? "no_role"} len=${tok.length}`);
    process.exit(payload.role === "service_role" ? 0 : 1);
  } catch {
    /* next */
  }
}
console.log(`role=decode_error len=${tok.length}`);
process.exit(1);

File: infra/hetzner/phase2-tail/verify-staging-key-role.py
#!/usr/bin/env python3
"""Print JWT role + length only — never the key."""
import base64
import json
import re
import sys

path = sys.argv[1] if len(sys.argv) > 1 else "/opt/souq-arab/config/api.env.staging"
text = open(path, encoding="utf-8").read()
m = re.search(r"^SUPABASE_SERVICE_ROLE_KEY=(.*)$", text, re.M)
if not m:
    print("role=missing len=0")
    sys.exit(1)
tok = m.group(1).strip().strip('"').strip("'")
parts = tok.split(".")
if len(parts) < 2:
    print(f"role=invalid_jwt len={len(tok)}")
    sys.exit(1)
payload = parts[1]
pad = "=" * (-len(payload) % 4)
raw = payload.replace("-", "+").replace("_", "/") + pad
try:
    data = json.loads(base64.b64decode(raw))
    role = data.get("role", "no_role")
    print(f"role={role} len={len(tok)}")
    sys.exit(0 if role == "service_role" else 1)
except Exception:
    print(f"role=decode_error len={len(tok)}")
    sys.exit(1)
